src/main/java/io/penguinstats/util/validator/DropsValidator.java

```diff
@@ -0,0 +1,98 @@
+package io.penguinstats.util.validator;
+
+import static java.util.stream.Collectors.groupingBy;
+import static java.util.stream.Collectors.toMap;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.springframework.stereotype.Component;
+
+import io.penguinstats.enums.DropType;
+import io.penguinstats.model.DropInfo;
+import io.penguinstats.model.TypedDrop;
+import io.penguinstats.service.DropInfoService;
+
+@Component("dropsValidator")
+public class DropsValidator extends BaseValidator {
+
+	private static Logger logger = LogManager.getLogger(DropsValidator.class);
+
+	private DropInfoService dropInfoService;
+
+	public DropsValidator(ValidatorContext context, DropInfoService dropInfoService) {
+		super(context);
+		this.dropInfoService = dropInfoService;
+	}
+
+	@Override
+	public boolean validate() {
+		List<TypedDrop> drops = this.context.getDrops();
+		if (drops == null)
+			return false;
+
+		Map<String, List<DropInfo>> openingDropInfosMap =
+				dropInfoService.getOpeningDropInfosMap(context.getServer(), context.getTimestamp());
+
+		if (!openingDropInfosMap.containsKey(context.getStageId())) // the stage was not open at the given timestamp
+			return false;
+
+		List<DropInfo> allDropInfos = openingDropInfosMap.get(context.getStageId());
+		Map<DropType, List<DropInfo>> dropInfoMapByDropType =
+				allDropInfos.stream().collect(groupingBy(DropInfo::getDropType));
+		Map<DropType, List<TypedDrop>> typedDropMapByDropType =
+				drops.stream().collect(groupingBy(TypedDrop::getDropType));
+
+		for (DropType dropType : DropType.values()) {
+			if (!dropInfoMapByDropType.containsKey(dropType)) {
+				logger.warn("Failed to find " + dropType + " drop info.");
+				continue;
+			}
+
+			List<TypedDrop> typedDrops = typedDropMapByDropType.getOrDefault(dropType, new ArrayList<>());
+			Map<String, TypedDrop> typedDropMapByItemId = null;
```

Comment:
Variable `typedDropMapByItemId` initializer `null` is redundant

Reply:
fixed